In [1]:
import pandas as pd
import numpy as np

In [2]:
nba_salaries=pd.read_csv(r'E:\mtech AI\Data Science\Completed Assignment\Chapter 6\nbasalaries.csv')
nba_salaries

PLAYER POSITION                TEAM  '15-'16 SALARY
0       Paul Millsap       PF       Atlanta Hawks       18.671659
1         Al Horford        C       Atlanta Hawks       12.000000
2     Tiago Splitter        C       Atlanta Hawks        9.756250
3        Jeff Teague       PG       Atlanta Hawks        8.000000
4        Kyle Korver       SG       Atlanta Hawks        5.746479
..               ...      ...                 ...             ...
412        Gary Neal       PG  Washington Wizards        2.139000
413     DeJuan Blair        C  Washington Wizards        2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards        1.920240
415   Garrett Temple       SG  Washington Wizards        1.100602
416     Jarell Eddie       SG  Washington Wizards        0.561716

[417 rows x 4 columns]

## Selecting Rows

Often, we would like to extract just those rows that correspond to entries with a particular feature. For example, we might want only the rows corresponding to the Warriors, or to players who earned more than  $10  million. Or we might just want the top five earners.



Specified Rows The Table method take does just that – it takes a specified set of rows. Its argument is a row index or array of indices, and it creates a new table consisting of only those rows.

For example, if we wanted just the first row of nba, we could use take as follows.

In [3]:
nba_salaries.iloc[[1]]

PLAYER POSITION           TEAM  '15-'16 SALARY
1  Al Horford        C  Atlanta Hawks            12.0

We could also get the fourth, fifth, and sixth rows by specifying a range of indices as the argument.

In [4]:
nba_salaries.iloc[[1,5,6]]

PLAYER POSITION           TEAM  '15-'16 SALARY
1       Al Horford        C  Atlanta Hawks       12.000000
5  Thabo Sefolosha       SF  Atlanta Hawks        4.000000
6       Mike Scott       PF  Atlanta Hawks        3.333333

In [5]:
nba_salaries['TEAM']

0           Atlanta Hawks
1           Atlanta Hawks
2           Atlanta Hawks
3           Atlanta Hawks
4           Atlanta Hawks
              ...        
412    Washington Wizards
413    Washington Wizards
414    Washington Wizards
415    Washington Wizards
416    Washington Wizards
Name: TEAM, Length: 417, dtype: object

If we want a table of the top 5 highest paid players, we can first sort the list by salary and then take the first five rows:

In [6]:
n=nba_salaries.sort_values(by=["'15-'16 SALARY"], ascending=True)
n.head(5)

PLAYER POSITION             TEAM  '15-'16 SALARY
267  Thanasis Antetokounmpo       SF  New York Knicks        0.030888
327          Cory Jefferson       PF     Phoenix Suns        0.049709
326            Jordan McRae       SG     Phoenix Suns        0.049709
324         Orlando Johnson       SG     Phoenix Suns        0.055722
325            Phil Pressey       PG     Phoenix Suns        0.055722

Rows Corresponding to a Specified Feature 
Rows Corresponding to a Specified Feature 
More often, we will want to access data in a set of rows that have a certain feature, but whose indices we don't know ahead of time. For example, we might want data on all the players who made more than  $10  million, but we don't want to spend time counting rows in the sorted table.The method where does the job for us. Its output is a table with the same columns as the original but only the rows where the feature occurs.

The first argument of where is the label of the column that contains the information about whether or not a row has the feature we want. If the feature is "made more than  $10  million", the column is SALARY.

The second argument of where is a way of specifying the feature. A couple of examples will make the general method of specification easier to understand.

In the first example, we extract the data for all those who earned more than  $10  million.

In [7]:
n.shape

(417, 4)

In [8]:
nba_salaries

PLAYER POSITION                TEAM  '15-'16 SALARY
0       Paul Millsap       PF       Atlanta Hawks       18.671659
1         Al Horford        C       Atlanta Hawks       12.000000
2     Tiago Splitter        C       Atlanta Hawks        9.756250
3        Jeff Teague       PG       Atlanta Hawks        8.000000
4        Kyle Korver       SG       Atlanta Hawks        5.746479
..               ...      ...                 ...             ...
412        Gary Neal       PG  Washington Wizards        2.139000
413     DeJuan Blair        C  Washington Wizards        2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards        1.920240
415   Garrett Temple       SG  Washington Wizards        1.100602
416     Jarell Eddie       SG  Washington Wizards        0.561716

[417 rows x 4 columns]

In [9]:
n=nba_salaries.sort_values(by=["'15-'16 SALARY"], ascending=False)
n

PLAYER POSITION                 TEAM  '15-'16 SALARY
169             Kobe Bryant       SF   Los Angeles Lakers       25.000000
29              Joe Johnson       SF        Brooklyn Nets       24.894863
72             LeBron James       SF  Cleveland Cavaliers       22.970500
255         Carmelo Anthony       SF      New York Knicks       22.875000
131           Dwight Howard        C      Houston Rockets       22.359364
..                      ...      ...                  ...             ...
200         Elliot Williams       SG    Memphis Grizzlies        0.055722
324         Orlando Johnson       SG         Phoenix Suns        0.055722
327          Cory Jefferson       PF         Phoenix Suns        0.049709
326            Jordan McRae       SG         Phoenix Suns        0.049709
267  Thanasis Antetokounmpo       SF      New York Knicks        0.030888

[417 rows x 4 columns]

In [10]:
sa=n["'15-'16 SALARY"]
sa
filter=sa>10
filter

169     True
29      True
72      True
255     True
131     True
       ...  
200    False
324    False
327    False
326    False
267    False
Name: '15-'16 SALARY, Length: 417, dtype: bool

In [11]:

n.where(filter, inplace = True)   
n

PLAYER POSITION                 TEAM  '15-'16 SALARY
169      Kobe Bryant       SF   Los Angeles Lakers       25.000000
29       Joe Johnson       SF        Brooklyn Nets       24.894863
72      LeBron James       SF  Cleveland Cavaliers       22.970500
255  Carmelo Anthony       SF      New York Knicks       22.875000
131    Dwight Howard        C      Houston Rockets       22.359364
..               ...      ...                  ...             ...
200              NaN      NaN                  NaN             NaN
324              NaN      NaN                  NaN             NaN
327              NaN      NaN                  NaN             NaN
326              NaN      NaN                  NaN             NaN
267              NaN      NaN                  NaN             NaN

[417 rows x 4 columns]

There are 69 rows in the new table, corresponding to the 69 players who made more than  10  million dollars. Arranging these rows in order makes the data easier to analyze. DeMar DeRozan of the Toronto Raptors was the "poorest" of this group, at a salary of just over  10  million dollars.

In [12]:
n1=n.sort_values(by=["'15-'16 SALARY"], ascending=True)
n1

PLAYER POSITION                TEAM  '15-'16 SALARY
368    DeMar DeRozan       SG     Toronto Raptors       10.050000
298   Gerald Wallace       SF  Philadelphia 76ers       10.105855
204        Luol Deng       SF          Miami Heat       10.151612
144      Monta Ellis       SG      Indiana Pacers       10.300000
95   Wilson Chandler       SF      Denver Nuggets       10.449438
..               ...      ...                 ...             ...
200              NaN      NaN                 NaN             NaN
324              NaN      NaN                 NaN             NaN
327              NaN      NaN                 NaN             NaN
326              NaN      NaN                 NaN             NaN
267              NaN      NaN                 NaN             NaN

[417 rows x 4 columns]

In [13]:
nba_salaries_filtered = nba_salaries[nba_salaries["'15-'16 SALARY"] >= 10]

In [14]:
print(nba_salaries_filtered.head(15)) 

              PLAYER POSITION                 TEAM  '15-'16 SALARY
0       Paul Millsap       PF        Atlanta Hawks       18.671659
1         Al Horford        C        Atlanta Hawks       12.000000
29       Joe Johnson       SF        Brooklyn Nets       24.894863
30    Thaddeus Young       PF        Brooklyn Nets       11.235955
42      Al Jefferson        C    Charlotte Hornets       13.500000
43     Nicolas Batum       SG    Charlotte Hornets       13.125306
44      Kemba Walker       PG    Charlotte Hornets       12.000000
60      Derrick Rose       PG        Chicago Bulls       20.093064
61      Jimmy Butler       SG        Chicago Bulls       16.407500
62       Joakim Noah        C        Chicago Bulls       13.400000
72      LeBron James       SF  Cleveland Cavaliers       22.970500
73        Kevin Love       PF  Cleveland Cavaliers       19.689000
74      Kyrie Irving       PG  Cleveland Cavaliers       16.407501
75  Tristan Thompson        C  Cleveland Cavaliers       14.26

How much did Stephen Curry make? For the answer, we have to access the row where the value of PLAYER is equal to Stephen Curry. That is placed a table consisting of just one line:

In [15]:
nba_salaries_filtered = nba_salaries[nba_salaries["PLAYER"] == 'Stephen Curry']

In [16]:
nba_salaries_filtered

PLAYER POSITION                   TEAM  '15-'16 SALARY
121  Stephen Curry       PG  Golden State Warriors       11.370786

Curry made just under  $11.4  million dollars. That's a lot of money, but it's less than half the salary of LeBron James. You'll find that salary in the "Top 5" table earlier in this section, or you could find it replacing 'Stephen Curry' by 'LeBron James' in the line of code above.

In the code, are is used again, but this time with the predicate equal_to instead of above. Thus for example you can get a table of all the Warriors:

In [17]:
nba_salaries_filtered = nba_salaries[nba_salaries["TEAM"] == 'Golden State Warriors']
nba_salaries_filtered

PLAYER POSITION                   TEAM  '15-'16 SALARY
117      Klay Thompson       SG  Golden State Warriors       15.501000
118     Draymond Green       PF  Golden State Warriors       14.260870
119       Andrew Bogut        C  Golden State Warriors       13.800000
120     Andre Iguodala       SF  Golden State Warriors       11.710456
121      Stephen Curry       PG  Golden State Warriors       11.370786
122     Jason Thompson       PF  Golden State Warriors        7.008475
123   Shaun Livingston       PG  Golden State Warriors        5.543725
124    Harrison Barnes       SF  Golden State Warriors        3.873398
125  Marreese Speights        C  Golden State Warriors        3.815000
126    Leandro Barbosa       SG  Golden State Warriors        2.500000
127       Festus Ezeli        C  Golden State Warriors        2.008748
128       Brandon Rush       SF  Golden State Warriors        1.270964
129       Kevon Looney       SF  Golden State Warriors        1.131960
130   Anderson Varejao       PF  Golden State Warriors        0.289755

This portion of the table is already sorted by salary, because the original table listed players sorted by salary within the same team. The .show() at the end of the line ensures that all rows are shown, not just the first 10.

It is so common to ask for the rows for which some column is equal to some value that the are.equal_to call is optional. Instead, the where method can be called with only a column name and a value to achieve the same effect.

In [18]:
filter1=nba_salaries['TEAM']
filter2=nba_salaries[nba_salaries["TEAM"] == 'Golden State Warriors']
filter2

PLAYER POSITION                   TEAM  '15-'16 SALARY
117      Klay Thompson       SG  Golden State Warriors       15.501000
118     Draymond Green       PF  Golden State Warriors       14.260870
119       Andrew Bogut        C  Golden State Warriors       13.800000
120     Andre Iguodala       SF  Golden State Warriors       11.710456
121      Stephen Curry       PG  Golden State Warriors       11.370786
122     Jason Thompson       PF  Golden State Warriors        7.008475
123   Shaun Livingston       PG  Golden State Warriors        5.543725
124    Harrison Barnes       SF  Golden State Warriors        3.873398
125  Marreese Speights        C  Golden State Warriors        3.815000
126    Leandro Barbosa       SG  Golden State Warriors        2.500000
127       Festus Ezeli        C  Golden State Warriors        2.008748
128       Brandon Rush       SF  Golden State Warriors        1.270964
129       Kevon Looney       SF  Golden State Warriors        1.131960
130   Anderson Varejao       PF  Golden State Warriors        0.289755

In [19]:
p=nba_salaries


In [20]:
p [p.TEAM=='Golden State Warriors']


PLAYER POSITION                   TEAM  '15-'16 SALARY
117      Klay Thompson       SG  Golden State Warriors       15.501000
118     Draymond Green       PF  Golden State Warriors       14.260870
119       Andrew Bogut        C  Golden State Warriors       13.800000
120     Andre Iguodala       SF  Golden State Warriors       11.710456
121      Stephen Curry       PG  Golden State Warriors       11.370786
122     Jason Thompson       PF  Golden State Warriors        7.008475
123   Shaun Livingston       PG  Golden State Warriors        5.543725
124    Harrison Barnes       SF  Golden State Warriors        3.873398
125  Marreese Speights        C  Golden State Warriors        3.815000
126    Leandro Barbosa       SG  Golden State Warriors        2.500000
127       Festus Ezeli        C  Golden State Warriors        2.008748
128       Brandon Rush       SF  Golden State Warriors        1.270964
129       Kevon Looney       SF  Golden State Warriors        1.131960
130   Anderson Varejao       PF  Golden State Warriors        0.289755

In [21]:
p.where(p.TEAM=='Golden State Warriors')   
p

PLAYER POSITION                TEAM  '15-'16 SALARY
0       Paul Millsap       PF       Atlanta Hawks       18.671659
1         Al Horford        C       Atlanta Hawks       12.000000
2     Tiago Splitter        C       Atlanta Hawks        9.756250
3        Jeff Teague       PG       Atlanta Hawks        8.000000
4        Kyle Korver       SG       Atlanta Hawks        5.746479
..               ...      ...                 ...             ...
412        Gary Neal       PG  Washington Wizards        2.139000
413     DeJuan Blair        C  Washington Wizards        2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards        1.920240
415   Garrett Temple       SG  Washington Wizards        1.100602
416     Jarell Eddie       SG  Washington Wizards        0.561716

[417 rows x 4 columns]

## Multiple Features

pYou can access rows that have multiple specified features, by using where repeatedly. For example, here is a way to extract all the Point Guards whose salaries were over  $15  million.

In [22]:
p.where(p.POSITION=='PG')   
p
p [p.TEAM=='Golden State Warriors']

PLAYER POSITION                   TEAM  '15-'16 SALARY
117      Klay Thompson       SG  Golden State Warriors       15.501000
118     Draymond Green       PF  Golden State Warriors       14.260870
119       Andrew Bogut        C  Golden State Warriors       13.800000
120     Andre Iguodala       SF  Golden State Warriors       11.710456
121      Stephen Curry       PG  Golden State Warriors       11.370786
122     Jason Thompson       PF  Golden State Warriors        7.008475
123   Shaun Livingston       PG  Golden State Warriors        5.543725
124    Harrison Barnes       SF  Golden State Warriors        3.873398
125  Marreese Speights        C  Golden State Warriors        3.815000
126    Leandro Barbosa       SG  Golden State Warriors        2.500000
127       Festus Ezeli        C  Golden State Warriors        2.008748
128       Brandon Rush       SF  Golden State Warriors        1.270964
129       Kevon Looney       SF  Golden State Warriors        1.131960
130   Anderson Varejao       PF  Golden State Warriors        0.289755

In [23]:
p.rename(columns = {"'15-'16 SALARY":'SALARY'}, inplace = True)
p

PLAYER POSITION                TEAM     SALARY
0       Paul Millsap       PF       Atlanta Hawks  18.671659
1         Al Horford        C       Atlanta Hawks  12.000000
2     Tiago Splitter        C       Atlanta Hawks   9.756250
3        Jeff Teague       PG       Atlanta Hawks   8.000000
4        Kyle Korver       SG       Atlanta Hawks   5.746479
..               ...      ...                 ...        ...
412        Gary Neal       PG  Washington Wizards   2.139000
413     DeJuan Blair        C  Washington Wizards   2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards   1.920240
415   Garrett Temple       SG  Washington Wizards   1.100602
416     Jarell Eddie       SG  Washington Wizards   0.561716

[417 rows x 4 columns]

In [24]:
p[(p.SALARY >= 15) & (p.POSITION == 'PG')]

PLAYER POSITION                   TEAM     SALARY
60        Derrick Rose       PG          Chicago Bulls  20.093064
74        Kyrie Irving       PG    Cleveland Cavaliers  16.407501
156         Chris Paul       PG   Los Angeles Clippers  21.468695
269  Russell Westbrook       PG  Oklahoma City Thunder  16.744218
400          John Wall       PG     Washington Wizards  15.851950

General Form
By now you will have realized that the general way to create a new table by selecting rows with a given feature is to use where and are with the appropriate condition

In [25]:
bool_series = p["SALARY"].between(10, 10.3, inclusive = True) 
p[bool_series]

PLAYER POSITION                TEAM     SALARY
144     Monta Ellis       SG      Indiana Pacers  10.300000
204       Luol Deng       SF          Miami Heat  10.151612
298  Gerald Wallace       SF  Philadelphia 76ers  10.105855
356     Danny Green       SG   San Antonio Spurs  10.000000
368   DeMar DeRozan       SG     Toronto Raptors  10.050000

If we specify a condition that isn't satisfied by any row, we get a table with column labels but no rows.

In [26]:
p[(p.SALARY >= 100) & (p.POSITION == 'PG')]

Empty DataFrame
Columns: [PLAYER, POSITION, TEAM, SALARY]
Index: []

## Some More Conditions 

The use of are.containing can help save some typing. For example, you can just specify Warriors instead of Golden State Warriors:

In [31]:
nba_salaries2 = nba_salaries[nba_salaries['TEAM'].str.contains("Warriors")] 


In [32]:
nba_salaries2

PLAYER POSITION                   TEAM     SALARY
117      Klay Thompson       SG  Golden State Warriors  15.501000
118     Draymond Green       PF  Golden State Warriors  14.260870
119       Andrew Bogut        C  Golden State Warriors  13.800000
120     Andre Iguodala       SF  Golden State Warriors  11.710456
121      Stephen Curry       PG  Golden State Warriors  11.370786
122     Jason Thompson       PF  Golden State Warriors   7.008475
123   Shaun Livingston       PG  Golden State Warriors   5.543725
124    Harrison Barnes       SF  Golden State Warriors   3.873398
125  Marreese Speights        C  Golden State Warriors   3.815000
126    Leandro Barbosa       SG  Golden State Warriors   2.500000
127       Festus Ezeli        C  Golden State Warriors   2.008748
128       Brandon Rush       SF  Golden State Warriors   1.270964
129       Kevon Looney       SF  Golden State Warriors   1.131960
130   Anderson Varejao       PF  Golden State Warriors   0.289755

You can extract data for all the guards, both Point Guards and Shooting Guards:

In [34]:
nba_salaries2 = nba_salaries[nba_salaries['POSITION'].str.contains("G")] 
nba_salaries2

PLAYER POSITION                TEAM    SALARY
3         Jeff Teague       PG       Atlanta Hawks  8.000000
4         Kyle Korver       SG       Atlanta Hawks  5.746479
8     Dennis Schroder       PG       Atlanta Hawks  1.763400
9    Tim Hardaway Jr.       SG       Atlanta Hawks  1.304520
11   Jason Richardson       SG       Atlanta Hawks  0.947276
..                ...      ...                 ...       ...
409     Alan Anderson       SG  Washington Wizards  4.000000
411    Ramon Sessions       PG  Washington Wizards  2.170465
412         Gary Neal       PG  Washington Wizards  2.139000
415    Garrett Temple       SG  Washington Wizards  1.100602
416      Jarell Eddie       SG  Washington Wizards  0.561716

[181 rows x 4 columns]

You can get all the players who were not Cleveland Cavaliers and had a salary of no less than  $20  million:

In [36]:
p[(p.SALARY >= 20) & (p.TEAM != 'Cleveland Cavaliers')]

PLAYER POSITION                   TEAM     SALARY
29       Joe Johnson       SF          Brooklyn Nets  24.894863
60      Derrick Rose       PG          Chicago Bulls  20.093064
131    Dwight Howard        C        Houston Rockets  22.359364
156       Chris Paul       PG   Los Angeles Clippers  21.468695
169      Kobe Bryant       SF     Los Angeles Lakers  25.000000
201       Chris Bosh       PF             Miami Heat  22.192730
202      Dwyane Wade       SG             Miami Heat  20.000000
255  Carmelo Anthony       SF        New York Knicks  22.875000
268     Kevin Durant       SF  Oklahoma City Thunder  20.158622